### PR TITLE
Update to Codemirror 6

### LINKED ISSUE #1635: Switch to CodeMirror 6
**Which part of the project should be enhanced?**
Markdown editor pane

**Is your enhancement request related to a problem? Please describe.**
Using CodeMirror 5 for the editor has historically led to various unfixable issues, like bad support for editing on mobile, screen readers or browser plugins that expect a working input field.

**Describe the solution you'd like**
[CodeMirror 6](https://codemirror.net/6/) is a rewrite of CodeMirror, which aims to fix all the aforementioned problems. We strive to use it in HedgeDoc 2.
After a quick investigation, I found these projects that provide a React component for CM6:
https://github.com/uiwjs/react-codemirror
https://github.com/sachinraja/rodemirror

**Describe alternatives you've considered**
Keep using CM5 with many issues.

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
CodeMirror 6

### Description
This PR upgrades CodeMirror to version 6

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] Added / updated documentation
- [x] Extended changelog
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### Things that don't work yet
- [ ] <del>Auto completion</del> will create issue
- [x] Code highlighting of other languages
- [x] Weird jump bug when using toolbar
- [ ] <del> Editor preferences?</del> will create issue
- [ ] <del> Drag'n'Drop of text still works text </del> will create issue

### Related Issue(s)
Fixes #1635

### CODE REVIEW

**Comment by mrdrogdrog** [comment]
I removed the auto completion and the preferences dialog for now.
Both components need a total rework. The auto completion sources should come from the extensions and the preferences dialog should be merged into the yet-to-create settings dialog.

File: src/components/render-page/iframe-markdown-renderer.tsx [hunk 1]
```diff
@@ -27,26 +27,50 @@ export const IframeMarkdownRenderer: React.FC = () => {
...
+    useCallback((values) => setFrontmatterInfo(values.frontmatterInfo), [])
+  )
+
+  useRendererReceiveHandler(
+    CommunicationMessageType.GET_WORD_COUNT,
+    useCallback(() => {
+      const documentContainer = document.querySelector('[data-word-count-target]')
+      communicator.sendMessageToOtherSide({
+        type: CommunicationMessageType.ON_WORD_COUNT_CALCULATED,
+        words: documentContainer ? countWords(documentContainer) : 0
```
Comment (ErikMichelson) [inline src/components/render-page/iframe-markdown-renderer.tsx:70]: Is the `countWords` method blocking the receive handler when operating on a huge content? Or should we wrap it inside a promise?

File: src/pages/n/[id].tsx [hunk 2]
```diff
@@ -64,22 +65,30 @@ export const EditorPage: NextPage = () => {
...
       if (scrollSource.current === ScrollSource.RENDERER && editorSyncScroll) {
         setScrollState((old) => {
-          const newState = { editorScrollState: newScrollState, rendererScrollState: old.rendererScrollState }
-          log.debug('Set scroll state because of renderer scroll', newState)
-          return newState
+          const newState: DualScrollState = {
+            editorScrollState: newScrollState,
+            rendererScrollState: old.rendererScrollState
+          }
+          return equal(newState, old) ? old : newState
```
Comment (ErikMichelson) [inline src/pages/n/[id].tsx:72]: If newState and old are equal why do you return the old state?
Always returning newState should be okay.

File: src/pages/n/[id].tsx [hunk 3]
```diff
@@ -64,22 +65,30 @@ export const EditorPage: NextPage = () => {
...
       if (scrollSource.current === ScrollSource.EDITOR && editorSyncScroll) {
         setScrollState((old) => {
-          const newState = { rendererScrollState: newScrollState, editorScrollState: old.editorScrollState }
-          log.debug('Set scroll state because of editor scroll', newState)
-          return newState
+          const newState: DualScrollState = {
+            rendererScrollState: newScrollState,
+            editorScrollState: old.editorScrollState
+          }
+          return equal(newState, old) ? old : newState
```
Comment (ErikMichelson) [inline src/pages/n/[id].tsx:91]: Same as above

File: src/redux/note-details/calculate-line-start-indexes.test.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { calculateLineStartIndexes } from './calculate-line-start-indexes'
+
+describe('calculateLineStartIndexes', () => {
+  it('works with an empty list', () => {
+    expect(calculateLineStartIndexes([])).toEqual([])
+  })
+  it('works with an non empty list', () => {
```
Comment (ErikMichelson) [inline src/redux/note-details/calculate-line-start-indexes.test.ts:13]: Maybe add a test case that empty lines work as well?
(Though they should work, it saves us of future mistakes)

**Review by ErikMichelson — COMMENTED** [review]
First batch

File: src/pages/n/[id].tsx (hunk 2 above)
Comment (mrdrogdrog) [reply src/pages/n/[id].tsx:72]: If I return new state then the returned object would be content equal but it's not same object instance.
By returning old state I suppress a rendering because of an state update.

File: src/pages/n/[id].tsx (hunk 3 above)
Comment (mrdrogdrog) [reply src/pages/n/[id].tsx:91]: Same as above.

File: src/components/editor-page/editor-pane/hooks/use-code-mirror-focus-reference.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import type { RefObject} from 'react';
+import { useMemo, useRef } from 'react'
+import { EditorView } from '@codemirror/view'
+import type { Extension } from '@codemirror/state'
+
+export const useCodeMirrorFocusReference = (): [Extension, RefObject<boolean>] => {
```
Comment (ErikMichelson) [inline src/components/editor-page/editor-pane/hooks/use-code-mirror-focus-reference.ts:12]: ESDoc missing

File: src/components/editor-page/editor-pane/hooks/use-off-screen-scroll-protection.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import { store } from '../../../../redux'
+import type { Extension } from '@codemirror/state'
+import { Logger } from '../../../../utils/logger'
+import type { ReactCodeMirrorRef } from '@uiw/react-codemirror'
+import { useNoteMarkdownContent } from '../../../../hooks/common/use-note-markdown-content'
+import { EditorView } from '@codemirror/view'
+
+const logger = new Logger('useOffScreenScrollProtection')
+
+export const useOffScreenScrollProtection = (codeMirrorRef: RefObject<ReactCodeMirrorRef>): Extension => {
```
Comment (ErikMichelson) [inline src/components/editor-page/editor-pane/hooks/use-off-screen-scroll-protection.ts:21]: ESDoc missing

File: src/redux/note-details/format-selection/apply-format-type-to-markdown-lines.test.ts [hunk 7]
```diff
@@ -7,46 +7,46 @@
 import { Mock } from 'ts-mockery'
 import * as wrapSelectionModule from './formatters/wrap-selection'
 import { applyFormatTypeToMarkdownLines } from './apply-format-type-to-markdown-lines'
-import type { CursorPosition, CursorSelection } from '../../editor/types'
+import type { CursorSelection } from '../../editor/types'
 import { FormatType } from '../types'
 import * as changeCursorsToWholeLineIfNoToCursorModule from './formatters/utils/change-cursors-to-whole-line-if-no-to-cursor'
-import * as replaceLinesOfSelectionModule from './formatters/replace-lines-of-selection'
+import * as replaceLinesOfSelectionModule from './formatters/prepend-lines-of-selection'
```
Comment (ErikMichelson) [inline src/redux/note-details/format-selection/apply-format-type-to-markdown-lines.test.ts:13]:
```suggestion
import * as prependLinesOfSelectionModule from './formatters/prepend-lines-of-selection'
```

File: src/redux/note-details/format-selection/apply-format-type-to-markdown-lines.test.ts [hunk 8]
```diff
@@ -58,180 +58,160 @@ describe('apply format type to markdown lines', () => {
...
   })
 
   it('can process the format type horizontal line with only from cursor', () => {
-    const fromCursor = Mock.of<CursorPosition>()
-    const result = applyFormatTypeToMarkdownLines(
-      markdownContentLinesMock,
-      { from: fromCursor },
-      FormatType.HORIZONTAL_LINE
-    )
+    const fromCursor = 138743857
```
Comment (ErikMichelson) [inline src/redux/note-details/format-selection/apply-format-type-to-markdown-lines.test.ts:139]: magic number or just some random value?

File: src/components/render-page/iframe-markdown-renderer.tsx (hunk 1 above)
Comment (mrdrogdrog) [reply src/components/render-page/iframe-markdown-renderer.tsx:70]: It shouldn't make any difference because every message is triggering a new event.

File: src/redux/note-details/format-selection/formatters/utils/change-cursors-to-whole-line-if-no-to-cursor.ts [hunk 9]
```diff
@@ -5,31 +5,63 @@
...
+
+  for (let characterIndex = adjustedStartPosition; characterIndex > 0; characterIndex -= 1) {
+    if (content.slice(characterIndex - 1, characterIndex) === '\n') {
+      return characterIndex
+    }
+  }
+  return 0
+}
+
+export const searchForEndOfLine = (content: string, startPosition: number): number => {
```
Comment (ErikMichelson) [inline src/redux/note-details/format-selection/formatters/utils/change-cursors-to-whole-line-if-no-to-cursor.ts:45]: ESDoc missing

File: src/redux/note-details/format-selection/formatters/utils/change-cursors-to-whole-line-if-no-to-cursor.ts [hunk 10]
```diff
@@ -5,31 +5,63 @@
...
+
+  for (let characterIndex = adjustedStartPosition; characterIndex < content.length; characterIndex += 1) {
+    if (content.slice(characterIndex, characterIndex + 1) === '\n') {
+      return characterIndex
+    }
+  }
+  return content.length
+}
+
+export const searchForNewline = (haystack: string, startPosition: number, step: number): number | undefined => {
```
Comment (ErikMichelson) [inline src/redux/note-details/format-selection/formatters/utils/change-cursors-to-whole-line-if-no-to-cursor.ts:56]: ESDoc missing

File: src/redux/note-details/format-selection/formatters/utils/change-cursors-to-whole-line-if-no-to-cursor.ts [hunk 11]
```diff
@@ -5,31 +5,63 @@
...
+  const newFrom = searchForStartOfLine(markdownContent, selection.from)
+  const newTo = searchForEndOfLine(markdownContent, selection.from)
+
+  return {
+    from: newFrom,
+    to: newTo
+  }
+}
+
+export const searchForStartOfLine = (content: string, startPosition: number): number => {
```
Comment (ErikMichelson) [inline src/redux/note-details/format-selection/formatters/utils/change-cursors-to-whole-line-if-no-to-cursor.ts:34]: ESDoc missing

**Review by ErikMichelson — COMMENTED** [review]
Batch 2

File: src/redux/note-details/format-selection/apply-format-type-to-markdown-lines.test.ts (hunk 8 above)
Comment (mrdrogdrog) [reply src/redux/note-details/format-selection/apply-format-type-to-markdown-lines.test.ts:139]: random. changed the name

File: src/components/render-page/iframe-markdown-renderer.tsx (hunk 1 above)
Comment (mrdrogdrog) [reply src/components/render-page/iframe-markdown-renderer.tsx:70]: I guess? @ErikMichelson ?

File: cypress/integration/fileUpload.spec.ts [hunk 12]
```diff
@@ -12,19 +12,6 @@ describe('File upload', () => {
     cy.fixture('demo.png').as('demoImage')
   })
 
-  it("doesn't prevent drag'n'drop of plain text", () => {
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply cypress/integration/fileUpload.spec.ts:15]: I tried. Really. But it doesn't work anymore.
And I have no idea how to fix this. Drag&Drop of text is an evil-bit-magic topic in cypress.

[the remark this answers is not in the document]
Comment (DerMolly) [reply cypress/integration/fileUpload.spec.ts:15]: Okay should we create an issue for this then?

File: src/components/editor-page/editor-pane/hooks/use-off-screen-scroll-protection.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+      },
+      focus: () => {
+        offFocusScrollState.current = undefined
+      }
+    })
+
+    const changeExtension = EditorView.updateListener.of((update) => {
+      const view = update.view
+      const scrollState = offFocusScrollState.current
+      if (!(!!scrollState && update.docChanged)) {
```
Comment (ErikMichelson) [inline src/components/editor-page/editor-pane/hooks/use-off-screen-scroll-protection.ts:41]: Simplify statement

**Review by ErikMichelson — APPROVED** [review]
